modules/python/python-persist.c

```diff
@@ -127,3 +168,330 @@ python_format_persist_name(const LogPipe *p, const gchar *module, PythonPersistM
+      PyErr_Format(PyExc_ValueError, "persist name cannot contain " SUBKEY_DELIMITER);
+      return -1;
+    }
+
+  if (!prepare_master_entry(self->persist_state, persist_name))
+    return -1;
+
+  if (!self->persist_name)
+    self->persist_name = g_strdup(persist_name);
+
+  return 0;
+}
+
+static void
+py_persist_dealloc(PyObject *s)
+{
+  PyPersist *self =(PyPersist *)s;
+
+  g_free(self->persist_name);
+  self->persist_name = NULL;
+
+  py_slng_generic_dealloc(s);
+}
+
+static PyMemberDef
+py_persist_type_members[] =
+{
+  {"persist_name", T_STRING, offsetof(PyPersist, persist_name), READONLY},
+  {NULL}
+};
+
+static gchar *
+_build_key(PyPersist *self, const gchar *key)
+{
+  return g_strdup_printf("%s" SUBKEY_DELIMITER "%s", self->persist_name, key);
+}
+
+static gchar *
+_lookup_entry(PyPersist *self, const gchar *key, guint8 *type)
+{
+  gchar *query_key = _build_key(self, key);
+
+  gsize size;
+  guint8 version;
+  PersistEntryHandle handle = persist_state_lookup_entry(self->persist_state, query_key, &size, &version);
+  if (!handle)
+    {
+      PyErr_Format(PyExc_KeyError, "Persist has no such key: %s", key);
+      g_free(query_key);
+      return NULL;
+    }
+
+  Entry *entry = persist_state_map_entry(self->persist_state, handle);
+  *type = entry->type;
+  gchar *copy = g_strdup(entry->data);
+  persist_state_unmap_entry(self->persist_state, handle);
+
+  g_free(query_key);
+
+  return copy;
+};
+
+static PersistEntryHandle
+_allocate_persist_entry(PersistState *persist_state, const gchar *key, const gchar *value, gsize value_len)
+{
+  gsize size;
+  guint8 version;
+  PersistEntryHandle handle = persist_state_lookup_entry(persist_state, key, &size, &version);
+  if (handle && size >= value_len)
+    return handle;
+
+  return persist_state_alloc_entry(persist_state, key, value_len);
+}
+
+static gchar *
+_serialize(guint8 type, PyObject *v)
+{
+  switch (type)
+    {
+    case ENTRY_TYPE_STRING:
+      return g_strdup(_py_get_string_as_string(v));
+    case ENTRY_TYPE_LONG:
+    {
+      PyObject *as_str = PyObject_Str(v);
+      g_assert(as_str);
+      gchar *result = g_strdup(_py_get_string_as_string(as_str));
+      Py_DECREF(as_str);
+      return result;
+    }
+    case ENTRY_TYPE_BYTES:
+      return g_strdup(PyBytes_AsString(v));
+    default:
+      g_assert_not_reached();
+    }
+}
+
+static gboolean
+_store_entry(PyPersist *self, const gchar *key, guint8 type, PyObject *v)
+{
+  gchar *query_key = _build_key(self, key);
+  gchar *value = _serialize(type, v);
+  gsize value_len = strlen(value) + sizeof(type);
+
+  PersistEntryHandle handle = _allocate_persist_entry(self->persist_state, query_key, value, value_len);
+  if (!handle)
+    {
+      g_free(value);
+      g_free(query_key);
+      return FALSE;
+    }
+
+  Entry *entry = persist_state_map_entry(self->persist_state, handle);
+  entry->type = type;
+  strcpy(entry->data, value);
+  persist_state_unmap_entry(self->persist_state, handle);
```

Comment:
I think we should assert on the fact that Entry does not have padding between type and value. I don't think there would be an issue ever with the current declaration however it is also a good way to document the assumption.

e.g. g_assert(sizeof(*entry) - sizeof(entry->data) == sizeof(entry->type));

Reply:
Thanks, fixed. I ended up with a `G_STATIC_ASSERT`